File: src/main/java/com/braydenoneal/blang/parser/expression/value/ItemValue.java
package com.braydenoneal.blang.parser.expression.value;

import com.mojang.serialization.MapCodec;
import com.mojang.serialization.codecs.RecordCodecBuilder;
import net.minecraft.item.Item;
import net.minecraft.registry.Registries;

public class ItemValue extends Value<Item> {
    public static final MapCodec<ItemValue> CODEC = RecordCodecBuilder.mapCodec(instance -> instance.group(
            Registries.ITEM.getCodec().fieldOf("value").forGetter(ItemValue::value)
    ).apply(instance, ItemValue::new));

    public ItemValue(Item value) {
        super(value);
    }

    @Override
    public ValueType<?> getValueType() {
        return ValueTypes.ITEM;
    }

    @Override
    public boolean equals(Object obj) {
        try {
            if (obj instanceof ItemValue itemValue) {
                return value().equals(itemValue.value());
            }
        } catch (Error e) {
            System.out.println("Cannot equate item values outside of the game");
        }

        return false;
    }
}

File: src/main/kotlin/com/braydenoneal/blang/parser/expression/value/ItemValue.kt
package com.braydenoneal.blang.parser.expression.value

import com.braydenoneal.blang.parser.RunException
import com.mojang.serialization.MapCodec
import com.mojang.serialization.codecs.RecordCodecBuilder
import net.minecraft.item.Item
import net.minecraft.registry.Registries

class ItemValue(value: Item) : Value<Item>(value) {
    override val valueType: ValueType<*> get() = ValueTypes.ITEM

    override fun equals(other: Any?): Boolean {
        try {
            if (other is ItemValue) {
                return value == other.value
            }
        } catch (_: Error) {
            throw RunException("Cannot equate item values outside of the game")
        }

        return false
    }

    companion object {
        val CODEC: MapCodec<ItemValue> = RecordCodecBuilder.mapCodec {
            it.group(
                Registries.ITEM.getCodec().fieldOf("value").forGetter(ItemValue::value)
            ).apply(it, ::ItemValue)
        }
    }

    override fun hashCode(): Int {
        var result = super.hashCode()
        result = 31 * result + valueType.hashCode()
        return result
    }
}
